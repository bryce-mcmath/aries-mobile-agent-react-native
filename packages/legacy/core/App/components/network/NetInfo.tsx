import { useEffect, useCallback, useState } from 'react'
import { useTranslation } from 'react-i18next'
import Toast from 'react-native-toast-message'
import { useNetwork } from '../../contexts/network'
import { ToastType } from '../../components/toast/BaseToast'
import { TOKENS, useServices } from '../../container-api'

const NetInfo: React.FC = () => {
  const { silentAssertConnectedNetwork, assertInternetReachable, assertMediatorReachable } = useNetwork()
  const [{ disableMediatorCheck }] = useServices([TOKENS.CONFIG])
  const { t } = useTranslation()
  const [hasShown, setHasShown] = useState(false)

  const showNetworkWarning = useCallback(() => {
    setHasShown(true)
    Toast.show({
      type: ToastType.Error,
      autoHide: true,
      text1: t('NetInfo.NoInternetConnectionTitle'),
    })
  }, [t])

  // will be null until network state is known
  const isConnected = silentAssertConnectedNetwork()

  useEffect(() => {
    // Only check general internet connection if mediator check is disabled
    if (disableMediatorCheck) {
      const internetReachable = assertInternetReachable()
      if (internetReachable) {
        Toast.hide()
      }
      
      // Strict check for false, null means the network state is not yet known
      if (internetReachable === false) {
        showNetworkWarning()
      }
      return
    }

    // Network is checked and available
    if (isConnected) {
      assertMediatorReachable().then((mediatorReachable) => {
        if (mediatorReachable) {
          Toast.hide()
          return
        }
        
        // Network is available but cannot access mediator, display toast
        showNetworkWarning()
      })
      return
    }
    
    // Network is checked and not connected, so display toast if not already shown
    if (isConnected === false && !hasShown) {
      showNetworkWarning()
    }
  }, [
    showNetworkWarning,
    isConnected,
    disableMediatorCheck,
    assertInternetReachable,
    assertMediatorReachable,
    hasShown
  ])

  return null
}

export default NetInfo
