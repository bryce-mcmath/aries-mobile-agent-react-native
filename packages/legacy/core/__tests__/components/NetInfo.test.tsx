import { render, waitFor } from '@testing-library/react-native'
import React from 'react'
import Toast from 'react-native-toast-message'
import { container } from 'tsyringe'

import mockNetworkContext from '../contexts/network'
import NetInfo from '../../App/components/network/NetInfo'
import toastConfig from '../../App/components/toast/ToastConfig'
import { CustomBasicAppContext } from '../helpers/app'
import { Container, TOKENS } from '../../App/container-api'
import { MainContainer } from '../../App/container-impl'

const createContainerContextForNetInfo = (disableMediatorCheck: boolean): Container => {
  const context = new MainContainer(container.createChildContainer()).init()
  const config = context.resolve(TOKENS.CONFIG)
  context.container.registerInstance(TOKENS.CONFIG, { ...config, disableMediatorCheck })
  return context
}

// Bifold's top offset for toasts
const topOffset = 15

describe('NetInfo Component', () => {
  describe('with mediator check enabled', () => {
    beforeEach(() => {
      jest.resetAllMocks()
    })

    const context = createContainerContextForNetInfo(false)

    it('should not show toast when network check has not finished', async () => {
      mockNetworkContext.silentAssertConnectedNetwork.mockReturnValue(null)
  
      const { queryByText } = render(
        <CustomBasicAppContext container={context}>
          <NetInfo />
          <Toast topOffset={topOffset} config={toastConfig} />
        </CustomBasicAppContext>
      )
  
      await waitFor(async () => {
        const toast = await queryByText('NetInfo.NoInternetConnectionTitle', { exact: false })
        expect(toast).toBeNull()
      })
    })

    test('should show toast when network unavailable', async () => {
      mockNetworkContext.silentAssertConnectedNetwork.mockReturnValue(false)
  
      const tree = render(
        <CustomBasicAppContext container={context}>
          <NetInfo />
          <Toast topOffset={topOffset} config={toastConfig} />
        </CustomBasicAppContext>
      )

      expect(tree).toMatchSnapshot()
  
      await waitFor(async () => {
        const toast = await tree.queryByText('NetInfo.NoInternetConnectionTitle', { exact: false })
        expect(toast).toBeTruthy()
      })
    })
  
    it('should show toast when network is connected but mediator is unreachable', async () => {
      mockNetworkContext.silentAssertConnectedNetwork.mockReturnValue(true)
      mockNetworkContext.assertMediatorReachable.mockResolvedValue(new Promise((resolve) => resolve(false)))
  
      const { queryByText } = render(
        <CustomBasicAppContext container={context}>
          <NetInfo />
          <Toast topOffset={topOffset} config={toastConfig} />
        </CustomBasicAppContext>
      )
  
      await waitFor(async () => {
        const toast = await queryByText('NetInfo.NoInternetConnectionTitle', { exact: false })
        expect(toast).toBeTruthy()
      })
    })
  
    it('should not show toast when network is connected and mediator is reachable', async () => {
      mockNetworkContext.silentAssertConnectedNetwork.mockReturnValue(true)
      mockNetworkContext.assertMediatorReachable.mockResolvedValue(new Promise((resolve) => resolve(true)))
  
      const { queryByText } = render(
        <CustomBasicAppContext container={context}>
          <NetInfo />
          <Toast topOffset={topOffset} config={toastConfig} />
        </CustomBasicAppContext>
      )
  
      await waitFor(async () => {
        const toast = await queryByText('NetInfo.NoInternetConnectionTitle', { exact: false })
        expect(toast).toBeNull()
      })
    })
  })

  describe('with mediator check disabled', () => {
    beforeEach(() => {
      jest.resetAllMocks()
    })

    const context = createContainerContextForNetInfo(true)

    it('should not show toast when internet is reachable', async () => {
      mockNetworkContext.assertInternetReachable.mockReturnValue(true)

      const { queryByText } = render(
        <CustomBasicAppContext container={context}>
          <NetInfo />
          <Toast topOffset={topOffset} config={toastConfig} />
        </CustomBasicAppContext>
      )
  
      await waitFor(async () => {
        const toast = await queryByText('NetInfo.NoInternetConnectionTitle', { exact: false })
        expect(toast).toBeNull()
      })
    })

    it('should show toast when internet is not reachable', async () => {
      mockNetworkContext.assertInternetReachable.mockReturnValue(false)

      const { queryByText } = render(
        <CustomBasicAppContext container={context}>
          <NetInfo />
          <Toast topOffset={topOffset} config={toastConfig} />
        </CustomBasicAppContext>
      )
  
      await waitFor(async () => {
        const toast = await queryByText('NetInfo.NoInternetConnectionTitle', { exact: false })
        expect(toast).toBeTruthy()
      })
    })

    it('should not show toast when internet reachability is unknown', async () => {
      mockNetworkContext.assertInternetReachable.mockReturnValue(null)

      const { queryByText } = render(
        <CustomBasicAppContext container={context}>
          <NetInfo />
          <Toast topOffset={topOffset} config={toastConfig} />
        </CustomBasicAppContext>
      )
  
      await waitFor(async () => {
        const toast = await queryByText('NetInfo.NoInternetConnectionTitle', { exact: false })
        expect(toast).toBeNull()
      })
    })
  })
})