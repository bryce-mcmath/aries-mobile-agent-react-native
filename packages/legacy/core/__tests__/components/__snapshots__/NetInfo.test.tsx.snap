// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NetInfo Component with mediator check enabled should show toast when network unavailable 1`] = `
<View
  collapsable={false}
  onLayout={[Function]}
  onMoveShouldSetResponder={[Function]}
  onMoveShouldSetResponderCapture={[Function]}
  onResponderEnd={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderReject={[Function]}
  onResponderRelease={[Function]}
  onResponderStart={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  onStartShouldSetResponderCapture={[Function]}
  pointerEvents="box-none"
  style={
    Object {
      "alignItems": "center",
      "justifyContent": "center",
      "left": 0,
      "opacity": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
      "transform": Array [
        Object {
          "translateY": 0,
        },
      ],
    }
  }
  testID="toastAnimatedContainer"
>
  <View
    accessibilityState={
      Object {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      Object {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "flex-start",
            "borderRadius": 4,
            "borderWidth": 2,
            "flex": 1,
            "flexDirection": "row",
            "marginTop": 25,
          },
          Object {
            "backgroundColor": "#313132",
            "borderColor": "#D8292F",
            "width": 675,
          },
        ]
      }
    >
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <Text
          allowFontScaling={false}
          selectable={false}
          style={
            Array [
              Object {
                "color": "#D8292F",
                "fontSize": 24,
              },
              Object {
                "marginHorizontal": 15,
                "marginTop": 15,
              },
              Object {
                "fontFamily": "Material Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          
        </Text>
        <View
          style={
            Object {
              "flexShrink": 1,
              "marginRight": 10,
              "marginVertical": 15,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#FFFFFF",
                  "fontSize": 18,
                  "fontWeight": "normal",
                },
                Object {
                  "fontWeight": "bold",
                },
                Object {
                  "color": "#FFFFFF",
                },
              ]
            }
            testID="com.ariesbifold:id/ToastTitle"
          >
            NetInfo.NoInternetConnectionTitle
          </Text>
        </View>
      </View>
      <View>
        <View
          accessibilityState={
            Object {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            Object {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <Text
            allowFontScaling={false}
            selectable={false}
            style={
              Array [
                Object {
                  "color": "#D8292F",
                  "fontSize": 24,
                },
                Object {
                  "marginHorizontal": 15,
                  "marginTop": 15,
                },
                Object {
                  "fontFamily": "Material Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
